File: apps/api/app/api/routes/patients.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.api.deps import get_db, require_role
from app.models.clinician import Clinician
from app.models.patient import Patient
from app.schemas.patient import (
    PatientLookupRequest,
    PatientMatchResult,
    PatientOut,
    PatientSearchHit,
    PriorVisitOut,
)
from app.services import audit
from app.services.patient_matching import (
    DEFAULT_SEARCH_LIMIT,
    create_patient,
    match_patient,
    previous_signed_note,
    search_patients_by_name,
)

router = APIRouter(prefix="/patients", tags=["patients"])


@router.post("/match", response_model=PatientMatchResult)
def match(
    payload: PatientLookupRequest,
    db: Session = Depends(get_db),
    # RBAC (0.2): patient search/match is a doctor action, part of the
    # same recording/identity workflow as encounters.
    clinician: Clinician = Depends(require_role("doctor")),
) -> PatientMatchResult:
    """P0-6: exact match links silently; near match needs a one-tap
    confirmation (client calls POST /patients with the confirmed name if
    the doctor picks a candidate, or accepts the exact match directly);
    no match means the client should create a new record.

    Audited as a PHI read (Phase 4.2). It is easy to read this route as a
    write path — it is a POST, and the client usually follows it with one —
    but it decrypts and ranks patient names to answer, and on an exact or
    near match it *discloses* an existing patient's identity to the caller.
    That is exactly the shape of access P0-8 asks to be accountable for.
    """
    result = match_patient(db, payload.name, payload.birthdate)
    audit.record(
        db,
        actor_clinician_id=clinician.id,
        action="patient.match",
        entity_type="patient",
        # The matched patient when there is one; "*" when the answer was
        # "nobody" — which still read the directory to find out. The
        # submitted name and birthdate are deliberately not recorded: they
        # are PHI, and this table keeps what it is given for years (see
        # app/models/audit_log.py). `match_type` is not PHI and is what
        # makes the row interpretable later.
        entity_id=result.patient.id if result.patient is not None else "*",
        diff={"match_type": result.match_type, "candidate_count": len(result.candidates)},
    )
    return result


@router.post("", response_model=PatientOut, status_code=201)
def create(
    payload: PatientLookupRequest,
    db: Session = Depends(get_db),
    clinician: Clinician = Depends(require_role("doctor")),
) -> PatientOut:
    """Creates a new record (P0-6: "no match creates a new record with
    name + birthdate"). Callers should have already called /match and
    confirmed match_type == "none" — this route does not re-check, since
    the one-tap confirmation UX for a near match also lands here with the
    doctor having explicitly chosen "this is a new patient."
    """
    patient = create_patient(db, payload.name, payload.birthdate)
    audit.record(
        db,
        actor_clinician_id=clinician.id,
        action="patient.create",
        entity_type="patient",
        entity_id=patient.id,
    )
    return PatientOut.model_validate(patient)


@router.get("/search", response_model=list[PatientSearchHit])
def search(
    q: str = Query(min_length=1, max_length=200, description="Typed or dictated patient name"),
    limit: int = Query(DEFAULT_SEARCH_LIMIT, ge=1, le=50),
    db: Session = Depends(get_db),
    clinician: Clinician = Depends(require_role("doctor")),
) -> list[PatientSearchHit]:
    """Name-first fuzzy search (P0-6: "accepts a typed or dictated patient
    name and fuzzy-matches against the existing directory").

    Distinct from `POST /match`, which needs a birthdate and answers the
    *deduplication* question. This answers the earlier one: which patients
    might the doctor mean? Birthdate comes back with each hit so two people
    with similar names can be told apart, which is why it is stored
    unencrypted in the first place.

    Note the audit entry: this is a PHI read — it decrypts every patient
    name in the directory to rank them — so it is logged as one. Phase 4.2
    exists because reads are the ones developers forget.
    """
    hits = search_patients_by_name(db, q, limit=limit)
    audit.record(
        db,
        actor_clinician_id=clinician.id,
        action="patient.search",
        entity_type="patient",
        # No single entity: a search reads the whole directory. The query
        # itself is deliberately NOT recorded — a patient name in an audit
        # log is PHI in a table with a longer retention than the record it
        # describes (Phase 4.2's own heads-up).
        entity_id="*",
    )
    return hits


@router.get("/{patient_id}/prior-visit", response_model=PriorVisitOut | None)
def prior_visit(
    patient_id: str,
    exclude_encounter_id: str | None = Query(None),
    db: Session = Depends(get_db),
    clinician: Clinician = Depends(require_role("doctor")),
) -> PriorVisitOut | None:
    """The prior visit's assessment and plan (P0-5's longitudinal-context
    item), for the note review screen.

    Returns null rather than 404 when there is no prior visit: a first-time
    patient is the normal case, not an error, and a 404 would push the client
    into treating an ordinary state as a failure.
    """
    note = previous_signed_note(db, patient_id, exclude_encounter_id=exclude_encounter_id)
    if note is None:
        return None

    audit.record(
        db,
        actor_clinician_id=clinician.id,
        action="note.read.prior_visit",
        entity_type="note",
        entity_id=note.id,
    )
    return PriorVisitOut(
        note_id=note.id,
        encounter_id=note.encounter_id,
        assessment=note.assessment,
        plan=note.plan,
        signed_at=note.signed_at,
    )


@router.get("/{patient_id}", response_model=PatientOut)
def get_patient(
    patient_id: str,
    db: Session = Depends(get_db),
    clinician: Clinician = Depends(require_role("doctor")),
) -> PatientOut:
    """Resolves a patient id to a name (found by this redesign's completeness
    audit as a gap: NoteReview re-opens an encounter already linked to a
    patient and had no route to ask who that is, so the signing screen — the
    highest-stakes screen in the app — showed a truncated UUID instead of a
    name). Registered after `/search` on purpose: `/{patient_id}` would
    otherwise swallow that literal path.

    A PHI read like search and prior-visit above, so it is audited the same
    way.
    """
    patient = db.get(Patient, patient_id)
    if patient is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Patient not found")

    audit.record(
        db,
        actor_clinician_id=clinician.id,
        action="patient.read",
        entity_type="patient",
        entity_id=patient.id,
    )
    return PatientOut.model_validate(patient)

File: apps/web/src/lib/patients.ts
/**
 * Patient identity (P0-6).
 *
 * The three-way outcome the PRD specifies is the whole design:
 *
 *   exact match  -> link silently, no prompt
 *   near match   -> one-tap confirmation
 *   no match     -> offer create-new with name + birthdate
 *
 * Note where birthdate enters. Search takes a **name only**, because that
 * is what a doctor has when they start typing. Birthdate governs
 * *deduplication* — it is shown alongside every candidate so two similar
 * names can be told apart, and it is required to create a record, because
 * "dedup uses name + birthdate together, not name alone".
 */
import { api, OfflineError } from "../api/client";

export type PatientHit = {
  id: string;
  full_name: string;
  birthdate: string;
  score: number;
  match_type: "exact" | "near";
};

export type SearchOutcome =
  /** Exactly one exact hit: P0-6 says link silently, so no confirmation. */
  | { kind: "exact"; patient: PatientHit }
  /** Candidates the doctor must choose between, or confirm one of. */
  | { kind: "near"; candidates: PatientHit[] }
  /** Nothing plausible: offer create-new. */
  | { kind: "none" }
  | { kind: "error"; reason: string };

export async function searchPatients(query: string): Promise<SearchOutcome> {
  if (!query.trim()) return { kind: "none" };

  try {
    const { data, error, response } = await api.GET("/api/v1/patients/search", {
      params: { query: { q: query, limit: 10 } },
    });

    if (error || !data) {
      return {
        kind: "error",
        reason:
          response.status === 403
            ? "Your account cannot search patients."
            : "Could not search the patient directory.",
      };
    }

    const hits = data as PatientHit[];
    const exact = hits.filter((h) => h.match_type === "exact");

    // "Links silently" applies only when there is exactly ONE exact match.
    // Two patients with the identical name is precisely the case where
    // silence would attach the note to the wrong person, so it falls through
    // to confirmation instead.
    if (exact.length === 1) return { kind: "exact", patient: exact[0] };
    if (hits.length > 0) return { kind: "near", candidates: hits };
    return { kind: "none" };
  } catch (e) {
    return {
      kind: "error",
      reason:
        e instanceof OfflineError
          ? "No connection, so the patient directory cannot be searched. Recording is not blocked — the session lands in the loose-sessions tray."
          : "Could not search the patient directory.",
    };
  }
}

/**
 * Creates a patient. Requires birthdate, and the server re-runs
 * name+birthdate matching, so a race between two doctors creating the same
 * patient resolves to one record rather than two.
 */
export async function createPatient(
  fullName: string,
  birthdate: string,
): Promise<{ ok: true; id: string } | { ok: false; reason: string }> {
  try {
    const { data, error } = await api.POST("/api/v1/patients", {
      body: { name: fullName, birthdate },
    });
    if (error || !data) return { ok: false, reason: "Could not create the patient record." };
    return { ok: true, id: data.id };
  } catch {
    return { ok: false, reason: "Could not create the patient record while offline." };
  }
}

/** Links a loose session to a patient (P0-6's one-tap linking action). */
export async function linkEncounterToPatient(
  encounterId: string,
  patientId: string,
): Promise<boolean> {
  try {
    const { response } = await api.POST("/api/v1/encounters/{encounter_id}/link-patient", {
      params: { path: { encounter_id: encounterId } },
      body: { patient_id: patientId },
    });
    return response.ok;
  } catch {
    return false;
  }
}

/**
 * Resolves a patient id to a name (P0 gap found by `/impeccable critique`:
 * NoteReview re-opens an encounter that was already linked to a patient —
 * usually the common case, not the first-link one PatientPicker handles —
 * and had no route to ask who that is, so the signing screen showed a
 * truncated UUID instead of a name). Returns null rather than throwing so a
 * failed lookup degrades to "no name available" instead of blocking the
 * note, matching how `fetchPriorVisit` already treats its own failures.
 */
export async function fetchPatient(
  patientId: string,
): Promise<{ id: string; full_name: string; birthdate: string } | null> {
  try {
    const { data } = await api.GET("/api/v1/patients/{patient_id}", {
      params: { path: { patient_id: patientId } },
    });
    return data ?? null;
  } catch {
    return null;
  }
}

export type PriorVisit = {
  note_id: string;
  encounter_id: string;
  assessment: string;
  plan: string;
  signed_at: string;
};

/**
 * The prior visit's assessment and plan (P0-5). Returns null for a
 * first-time patient, which is an ordinary state rather than an error.
 */
export async function fetchPriorVisit(
  patientId: string,
  excludeEncounterId?: string,
): Promise<PriorVisit | null> {
  try {
    const { data } = await api.GET("/api/v1/patients/{patient_id}/prior-visit", {
      params: {
        path: { patient_id: patientId },
        query: excludeEncounterId ? { exclude_encounter_id: excludeEncounterId } : {},
      },
    });
    return (data as PriorVisit | null) ?? null;
  } catch {
    return null;
  }
}
